airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/AbstractJdbcSource.java

```diff
@@ -292,12 +298,57 @@ public AirbyteCatalog discover(JsonNode config) throws Exception {
               .map(f -> Field.of(f.getColumnName(), JdbcUtils.getType(f.getColumnType())))
               .distinct()
               .collect(Collectors.toList());
-
-          return new TableInfo(JdbcUtils.getFullyQualifiedTableName(t.getSchemaName(), t.getName()), fields, t.getPrimaryKeys());
+          final String streamName = JdbcUtils.getFullyQualifiedTableName(t.getSchemaName(), t.getName());
+          final List<String> primaryKeys;
+          if (tablePrimaryKeys.isEmpty()) {
+            // Some database can't return primary keys without specifying a target table name, retrying while
+            // being more specific
+            primaryKeys = discoverPrimaryKeys(database, databaseOptional, Optional.of(t.getSchemaName()), t.getName())
+                .getOrDefault(streamName, Collections.emptyList());
+          } else {
+            primaryKeys = tablePrimaryKeys.getOrDefault(streamName, Collections.emptyList());
+          }
+          return new TableInfo(streamName, fields, primaryKeys);
         })
         .collect(Collectors.toList());
   }
 
+  /**
+   * Discover Primary keys for each table and @return a map of schema.table name to their associated
+   * list of primary key fields.
+   *
+   * If @param tableName is not specified, this function may fail on some databases (for example
+   * MySql) but works on others (for instance Postgres)
+   */
+  private Map<String, List<String>> discoverPrimaryKeys(JdbcDatabase database,
```

Comment:
got it. i'd suggest 1 of 3 options here:
1. have this return an optional if the query doesn't work
2. have this method always succeed by hiding this logic internally. So it always takes the table names. If it doesn't need them it doesn't use them, but if it does it has them.
e.g. `discoverPrimaryKeys(JdbcDatabase database, Optional<String> databaseOptional, Optional<String> schemaOptional, String tableName)` => `discoverPrimaryKeys(JdbcDatabase database, Optional<String> databaseOptional, Optional<String> schemaOptional, List<String> tableNames)`
3. just always query one table at a time.

i think 2 or 3 are equally good options and are clearer than 1. with the weird quirk you discovered that you can't always use null for `getPrimaryKeys` it's not the end of the world if we just do a query per table.